core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -641,44 +716,78 @@ class LogManager(logDirs: Seq[File],
   }
 
   /**
+    * Mark the partition directory in the source log directory for deletion and
+    * rename the future log of this partition in the destination log directory to be the current log
+    *
+    * @param topicPartition TopicPartition that needs to be swapped
+    */
+  def replaceCurrentWithFutureLog(topicPartition: TopicPartition): Unit = {
+    logCreationOrDeletionLock synchronized {
+      val sourceLog = currentLogs.get(topicPartition)
+      val destLog = futureLogs.get(topicPartition)
+
+      if (sourceLog == null)
+        throw new KafkaStorageException(s"The current replica for $topicPartition is offline")
+      if (destLog == null)
+        throw new KafkaStorageException(s"The future replica for $topicPartition is offline")
+
+      destLog.renameDir(Log.logDirName(topicPartition))
+      // Now that future replica has been successfully renamed to be the current replica
+      // Update the cached map and log cleaner as appropriate.
+      futureLogs.remove(topicPartition)
+      currentLogs.put(topicPartition, destLog)
+      if (cleaner != null) {
+        cleaner.alterCheckpointDir(topicPartition, sourceLog.dir.getParentFile, destLog.dir.getParentFile)
+        cleaner.resumeCleaning(topicPartition)
+      }
+
+      try {
+        sourceLog.renameDir(Log.logDeleteDirName(topicPartition))
+        // Now that replica in source log directory has been successfully renamed for deletion.
+        // Close the log, update checkpoint files, and enqueue this log to be deleted.
+        sourceLog.close()
+        checkpointLogRecoveryOffsetsInDir(sourceLog.dir.getParentFile)
+        checkpointLogStartOffsetsInDir(sourceLog.dir.getParentFile)
+        logsToBeDeleted.add(sourceLog)
+      } catch {
+        case e: KafkaStorageException =>
+          // If sourceLog's log directory is offline, we need close its handlers here.
+          // handleLogDirFailure() will not close handlers of sourceLog because it has been removed from currentLogs map
+          sourceLog.closeHandlers()
```

Comment:
Hmm, it seems that sourceLog.close() can only throw KafkaStorageException because closeHandlers() has already been called due to offline log dir. So, not sure why closeHandlers() needs to be called again.

Reply:
I think Log.close() also does disk IO operation, e.g. when it calls producerStateManager.takeSnapshot() or FileRecords.flush(). Therefore it could throw KafkaStorageException() if the IOException occurs.

I realized that we need maybeHandleIOException() for Log.close() to catch e.g. IOException from producerStateManager.takeSnapshot(). I have updated the patch to do this.

Reply:
Thanks, sound goods.